### PR TITLE
Implement indexed access on select elements

### LINKED ISSUE #11763: Implement indexed access on select elements
This also includes the `length` attribute.

It is probably simpler not to involve the `select.options` collection for now, and just implement the getter directly.

* Spec: <https://html.spec.whatwg.org/multipage/#dom-select-item>.
* Code: `components/script/dom/htmlselectelement.rs`, `components/script/dom/webidls/HTMLSelectElement.rs`.
* Test: `/html/semantics/forms/the-select-element/common-HTMLOptionsCollection.html`, `FileAPI/blob/Blob-constructor.html`.

### PR DESCRIPTION
**Pull request by mskrzypkows**
Added methods for indexed access on select: SetLength, Length, Item, IndexedGetter

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11763 (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13066)

### CODE REVIEW

File: components/script/dom/htmlselectelement.rs [hunk 1]
```diff
@@ -183,6 +184,65 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
+                    match removed {
+                        Ok(_) => {},
+                        Err(e) => println!("error removing child of HTML Select: {:?}", e),
+                    }
+                } else {
+                    println!("error: there is no child of HTML Select" );
+                }
+            }
+        } else if value > length {  // add new blank option elements
+            let document = node.GetOwnerDocument();
```
Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: `document_from_node(self)`

File: components/script/dom/htmlselectelement.rs [hunk 2]
```diff
@@ -183,6 +184,65 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
+            if document.is_some() {
+                let document = document.unwrap();
+                for _ in 0..(value - length) {
+                    let element = HTMLOptionElement::new(atom!("option"), None, &document.upcast());
+                    let appended = node.AppendChild(element.upcast());
+                    match appended {
+                        Ok(_) => {},
+                        Err(e) => println!("error appending child of HTML Select: {:?}", e),
```
Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]:
```rust
if let Err(e) = node.AppendChild(element.upcast()) {
    println!("error appending child of HTMLSelectElement: {:?}", e);
}
```

File: components/script/dom/htmlselectelement.rs [hunk 3]
```diff
@@ -183,6 +184,65 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
+    fn Length(&self) -> u32 {
+        let node = self.upcast::<Node>();
+        node.traverse_preorder().filter_map(Root::downcast::<HTMLOptionElement>).count() as u32
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-item
+    fn Item(&self, index: u32) -> Option<Root<Element>> {
+        let node = self.upcast::<Node>();
+        let item = node.traverse_preorder().filter_map(Root::downcast::<HTMLOptionElement>).nth(index as usize);
+        if item.is_some() {
```
Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: ` if let Some(item) = item {`

Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: Actually, we can even do this:
```rust
item.map(|i| Root::from_ref(i.upcast()))
```

Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: Or, if you don't mind chaining up functions:
```rust
node.traverse_preorder()
    .filter_map(Root::downcast::<HTMLOptionElement>)
    .nth(index as usize)
    .map(|item| Root::from_ref(item.upcast()))
```

File: components/script/dom/htmlselectelement.rs [hunk 4]
```diff
@@ -183,6 +184,65 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
         self.upcast::<HTMLElement>().labels()
     }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-length
+    fn SetLength(&self, value: u32) {
+        let length  = self.Length();
+        let node = self.upcast::<Node>();
+        if value < length {     // truncate the number of option elements
+            for _ in 0..(length - value) {
+                let lastChild = node.GetLastChild();
```
Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: I think we can just use `node.descending_last_children().take(length - value)` here.

Comment (mskrzypkows) [inline components/script/dom/htmlselectelement.rs]: For some reason
```
for child in node.descending_last_children().take((length - value) as usize) {
    if let Err(e) = node.RemoveChild(&child) {
        println!("error removing child of HTML Select: {:?}", e);
    }
}
```
doesn't work. I'll check it and upload next commit.

Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]:
```rust
let iter = node.descending_last_children().take((length - value) as usize);
while let Some(child) = iter.next() {
    if let Err(e) = node.RemoveChild(&child) {
        println!("Error removing child of HTMLSelectElement: {:?}", e);
    }
}
```

**Comment by Manishearth** [comment]
Any updates? If you need help, just ask! 😄

**Comment by mskrzypkows** [comment]
I'll update today, sorry for the delay, I was busy recently.

**Comment by Manishearth** [comment]
No problem :)

File: components/script/dom/htmlselectelement.rs (hunk 4 above)
Comment (mskrzypkows) [inline components/script/dom/htmlselectelement.rs]: Still doesn't work, removing children test fails.

File: components/script/dom/htmlselectelement.rs [hunk 5]
```diff
@@ -183,6 +184,51 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
         self.upcast::<HTMLElement>().labels()
     }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-length
+    fn SetLength(&self, value: u32) {
+        let length  = self.Length();
+        let node = self.upcast::<Node>();
+        if value < length {     // truncate the number of option elements
+            for _ in 0..(length - value) {
+                if let Some(lastChild) = node.GetLastChild() {
```
Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: Looks like you ignored my review comments for this part.

File: components/script/dom/htmlselectelement.rs [hunk 6]
```diff
@@ -183,6 +184,51 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
...
+                if let Err(e) = node.AppendChild(element.upcast()) {
+                    warn!("error appending child of HTMLSelectElement: {:?}", e);
+                }
+            }
+        }
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-length
+    fn Length(&self) -> u32 {
+        let node = self.upcast::<Node>();
```
Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: This can be melded into the following line without a let-binding to node.

**Review by KiChjang — CHANGES_REQUESTED** [review]

File: components/script/dom/htmlselectelement.rs (hunk 5 above)
Comment (mskrzypkows) [inline components/script/dom/htmlselectelement.rs]: I didn't ignore your comment, check my reply to your previous comment. It doesn't work as you mentioned children nodes aren't deleted.

Comment (KiChjang) [inline components/script/dom/htmlselectelement.rs]: Sorry, `descending_last_children` is incorrect. Use `rev_children` instead.

File: tests/wpt/metadata/html/semantics/forms/the-select-element/common-HTMLOptionsCollection.html.ini [hunk 7]
```diff
@@ -1,18 +1,2 @@
 [common-HTMLOptionsCollection.html]
   type: testharness
```
Comment (KiChjang) [inline tests/wpt/metadata/html/semantics/forms/the-select-element/common-HTMLOptionsCollection.html.ini]: This file can be deleted.

**Review by KiChjang — CHANGES_REQUESTED** [review]

File: components/script/dom/htmlselectelement.rs (hunk 5 above)
Comment (mskrzypkows) [inline components/script/dom/htmlselectelement.rs]: OK, thanks for hints.

**Comment by KiChjang** [comment]
Squash!

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLSelectElement interface: attribute length

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLSelectElement interface: operation item(unsigned long)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLSelectElement interface: document.createElement("select") must inherit property "length" with the proper type (10)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLSelectElement interface: document.createElement("select") must inherit property "item" with the proper type (11)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLSelectElement interface: calling item(unsigned long) on document.createElement("select") with too few arguments must throw TypeError
```

**Comment by mskrzypkows** [comment]
Strange build error, it seems not related to my changes.

**Review by KiChjang — APPROVED** [review]

**Comment by KiChjang** [comment]
@bors-servo retry
* #12654, #13359

**Pull request merged**